utils/version/gen_version.py

```diff
@@ -0,0 +1,107 @@
+import argparse
```

Comment:
I'd like a header comment summarizing the version scheme implemented here. We have no documentation for it outside of the code itself, and while not too complex, it still needs reading through. This could be a place to write why we chose that scheme too, to leave a paper trail.

Reply:
Yes, I've just finished updating the PR description and I will copy that here.

Reply:
Perfect!